File: Models/book model/book_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class BookModel {
  final String id;
  final String title;
  final String description;
  final int estimatedReadTime;
  final int levelNumber;
  final String pdfUrl;
  final String accessTier;
  final bool isActive;
  final DateTime? publishedAt;
  final int sortOrder;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String createdBy;
  final int totalPages;

  const BookModel({
    required this.id,
    required this.title,
    required this.description,
    required this.estimatedReadTime,
    required this.levelNumber,
    required this.pdfUrl,
    required this.accessTier,
    required this.isActive,
    this.publishedAt,
    required this.sortOrder,
    required this.createdAt,
    required this.updatedAt,
    required this.createdBy,
    required this.totalPages, // Add to constructor
  });

  factory BookModel.create({
    required String id,
    required String title,
    required String description,
    required int estimatedReadTime,
    required int levelNumber,
    required String pdfUrl,
    String accessTier = 'free',
    required int totalPages, // Add to factory
    bool isActive = true,
    DateTime? publishedAt,
    int sortOrder = 0,
    String createdBy = '',
  }) {
    final now = DateTime.now();
    return BookModel(
      id: id,
      title: title,
      description: description,
      estimatedReadTime: estimatedReadTime,
      levelNumber: levelNumber,
      pdfUrl: pdfUrl,
      accessTier: accessTier,
      isActive: isActive,
      publishedAt: publishedAt,
      sortOrder: sortOrder,
      createdAt: now,
      updatedAt: now,
      createdBy: createdBy,
      totalPages: totalPages, // Initialize
    );
  }

  Map<String, dynamic> toFirestore() => {
    'id': id,
    'title': title,
    'description': description,
    'estimatedReadTime': estimatedReadTime,
    'levelNumber': levelNumber,
    'pdfUrl': pdfUrl,
    'accessTier': accessTier,
    'isActive': isActive,
    'publishedAt': publishedAt == null
        ? null
        : Timestamp.fromDate(publishedAt!),
    'sortOrder': sortOrder,
    'createdAt': Timestamp.fromDate(createdAt),
    'updatedAt': Timestamp.fromDate(updatedAt),
    'createdBy': createdBy,
    'totalPages': totalPages, // Add to Firestore map
  };

  factory BookModel.fromFirestore(Map<String, dynamic> json) => BookModel(
    id: json['id'] ?? '',
    title: json['title'] ?? '',
    description: json['description'] ?? '',
    estimatedReadTime: json['estimatedReadTime'] ?? 0,
    levelNumber: json['levelNumber'] ?? 1,
    pdfUrl: json['pdfUrl'] ?? '',
    accessTier: json['accessTier'] ?? 'free',
    isActive: json['isActive'] ?? true,
    publishedAt: _toDateTime(json['publishedAt']),
    sortOrder: json['sortOrder'] ?? 0,
    createdAt: _toDateTime(json['createdAt']),
    updatedAt: _toDateTime(json['updatedAt']),
    createdBy: json['createdBy'] ?? '',
    totalPages: json['totalPages'] ?? 0, // Parse totalPages
  );

  BookModel copyWith({
    String? id,
    String? title,
    String? description,
    int? estimatedReadTime,
    int? levelNumber,
    String? pdfUrl,
    String? accessTier,
    bool? isActive,
    DateTime? publishedAt,
    int? sortOrder,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? createdBy,
  }) => BookModel(
    id: id ?? this.id,
    title: title ?? this.title,
    description: description ?? this.description,
    estimatedReadTime: estimatedReadTime ?? this.estimatedReadTime,
    levelNumber: levelNumber ?? this.levelNumber,
    pdfUrl: pdfUrl ?? this.pdfUrl,
    accessTier: accessTier ?? this.accessTier,
    isActive: isActive ?? this.isActive,
    publishedAt: publishedAt ?? this.publishedAt,
    sortOrder: sortOrder ?? this.sortOrder,
    createdAt: createdAt ?? this.createdAt,
    updatedAt: updatedAt ?? this.updatedAt,
    createdBy: createdBy ?? this.createdBy,
    totalPages: totalPages ?? this.totalPages,
  );

  static DateTime _toDateTime(dynamic v) {
    if (v is Timestamp) return v.toDate();
    if (v is String) return DateTime.parse(v);
    return DateTime.now();
  }
}

File: src/model/BookModel.ts
import { Timestamp, DocumentData } from "firebase/firestore";

export interface IBookModel {
  id: string;
  title: string;
  description: string;
  // estimatedReadTime: number;
  levelNumber: number;
  pdfUrl: string;
  accessTier: "free" | "premium";
  isActive: boolean;
  publishedAt?: Date | null;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  totalPages: number;
  levelId: string; 
}

export class BookModel implements IBookModel {
  id: string;
  title: string;
  description: string;
  // estimatedReadTime: number;
  levelNumber: number;
  pdfUrl: string;
  accessTier: "free" | "premium";
  isActive: boolean;
  publishedAt?: Date | null;
  sortOrder: number;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
  totalPages: number;
  levelId: string;

  constructor(data: IBookModel) {
    this.id = data.id;
    this.title = data.title;
    this.description = data.description;
    // this.estimatedReadTime = data.estimatedReadTime;
    this.levelNumber = data.levelNumber;
    this.pdfUrl = data.pdfUrl;
    this.accessTier = data.accessTier;
    this.isActive = data.isActive;
    this.publishedAt = data.publishedAt ?? null;
    this.sortOrder = data.sortOrder;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.createdBy = data.createdBy;
    this.totalPages = data.totalPages;
    this.levelId = data.levelId;
  }

  static create(params: {
    id: string;
    title: string;
    description: string;
    // estimatedReadTime: number;
    levelNumber: number;
    pdfUrl: string;
    accessTier?: "free" | "premium";
    totalPages: number;
    isActive?: boolean;
    publishedAt?: Date | null;
    sortOrder?: number;
    createdBy?: string;
    levelId?: string;
  }): BookModel {
    const now = new Date();
    return new BookModel({
      id: params.id,
      title: params.title,
      description: params.description,
      // estimatedReadTime: params.estimatedReadTime,
      levelNumber: params.levelNumber,
      pdfUrl: params.pdfUrl,
      accessTier: params.accessTier ?? "free",
      totalPages: params.totalPages,
      isActive: params.isActive ?? true,
      publishedAt: params.publishedAt ?? null,
      sortOrder: params.sortOrder ?? 0,
      createdBy: params.createdBy ?? "admin",
      createdAt: now,
      updatedAt: now,
      levelId: params.levelId,
    });
  }

  static fromFirestore(data: DocumentData, idFromSnap?: string): BookModel {
    return new BookModel({
      id: idFromSnap ?? data.id ?? "",
      title: data.title ?? "",
      description: data.description ?? "",
      // estimatedReadTime: Number(data.estimatedReadTime ?? 0),
      levelNumber: Number(data.levelNumber ?? 1),
      pdfUrl: data.pdfUrl ?? "",
      accessTier: (data.accessTier ?? "free") as "free" | "premium",
      isActive: Boolean(data.isActive ?? true),
      publishedAt: BookModel._toDate(data.publishedAt),
      sortOrder: Number(data.sortOrder ?? 0),
      createdAt: BookModel._toDate(data.createdAt) ?? new Date(),
      updatedAt: BookModel._toDate(data.updatedAt) ?? new Date(),
      createdBy: data.createdBy ?? "admin",
      totalPages: Number(data.totalPages ?? 0),
      levelId: data.levelId ?? "",
    });
  }

  toFirestore(): any {
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      // estimatedReadTime: this.estimatedReadTime,
      levelNumber: this.levelNumber,
      pdfUrl: this.pdfUrl,
      accessTier: this.accessTier,
      isActive: this.isActive,
      publishedAt: this.publishedAt ? Timestamp.fromDate(this.publishedAt) : null,
      sortOrder: this.sortOrder,
      createdAt: Timestamp.fromDate(this.createdAt),
      updatedAt: Timestamp.fromDate(this.updatedAt),
      createdBy: this.createdBy,
      totalPages: this.totalPages,
      levelId: this.levelId,
    };
  }

  copyWith(params: Partial<IBookModel>): BookModel {
    return new BookModel({
      id: params.id ?? this.id,
      title: params.title ?? this.title,
      description: params.description ?? this.description,
      // estimatedReadTime: params.estimatedReadTime ?? this.estimatedReadTime,
      levelNumber: params.levelNumber ?? this.levelNumber,
      pdfUrl: params.pdfUrl ?? this.pdfUrl,
      accessTier: (params.accessTier ?? this.accessTier) as "free" | "premium",
      isActive: params.isActive ?? this.isActive,
      publishedAt: params.publishedAt ?? this.publishedAt,
      sortOrder: params.sortOrder ?? this.sortOrder,
      createdAt: params.createdAt ?? this.createdAt,
      updatedAt: params.updatedAt ?? this.updatedAt,
      createdBy: params.createdBy ?? this.createdBy,
      totalPages: params.totalPages ?? this.totalPages,
      levelId: params.levelId ?? this.levelId,
    });
  }

  private static _toDate(value: any): Date | null {
    if (!value) return null;
    if (value instanceof Timestamp) return value.toDate();
    if (value instanceof Date) return value;
    if (typeof value === "string") return new Date(value);
    return null;
  }
}
